python_onnxifi/data_conversion.cc

```diff
@@ -0,0 +1,251 @@
+  } 			       							 \
+  case ONNX_NAMESPACE::TensorProto_DataType_UINT8:  {				 \
+    op_template<int32_t, uint8_t>(__VA_ARGS__);          			 \
+    break;									 \
+  } 		 								 \
+  case ONNX_NAMESPACE::TensorProto_DataType_UINT16: { 				 \
+    op_template<int32_t, uint16_t>(__VA_ARGS__);          		    	 \
+    break;									 \
+  } 									         \
+  case ONNX_NAMESPACE::TensorProto_DataType_INT64: {				 \
+    op_template<int64_t, int64_t>(__VA_ARGS__);                               	 \
+    break;                                                               	 \
+  }										 \
+  case ONNX_NAMESPACE::TensorProto_DataType_UINT32:  {				 \
+    op_template<uint64_t, uint32_t>(__VA_ARGS__);                                \
+    break;                                                                       \
+  }     									 \
+  case ONNX_NAMESPACE::TensorProto_DataType_UINT64: {			 	 \
+    op_template<uint64_t, uint64_t>( __VA_ARGS__);                            	 \
+    break;							        	 \
+  }   						                        	 \
+  case ONNX_NAMESPACE::TensorProto_DataType_DOUBLE: {                   	 \
+   op_template<double, double>(__VA_ARGS__);                                 	 \
+   break;                                                                        \
+  }										 \
+  case ONNX_NAMESPACE::TensorProto_DataType_COMPLEX128: {     			 \
+   op_template<std::complex<double>, std::complex<double>>(__VA_ARGS__);         \
+   break;									 \
+  }										 \
+  case ONNX_NAMESPACE::TensorProto_DataType_STRING: 				 \
+  case ONNX_NAMESPACE::TensorProto_DataType_UNDEFINED: {			 \
+   throw std::runtime_error("Dispatch received non-numeric data type");          \
+  }										 \
+}										 \
+
+template<typename onnx_type, typename unused>
+void DataConversion::getBufferSize(uint64_t& buffer_size, const uint64_t* shape, uint32_t dimensions)  {
+  auto numElements = std::accumulate(shape, shape + dimensions,
+                                    (uint64_t) 1, std::multiplies<uint64_t>());
+  buffer_size = sizeof(onnx_type) * numElements;
+}
+
+void DataConversion::updateOutputDescriptors(ModelProto& model)  {
+  releaseDescriptors(output_descriptors_);
+  ::ONNX_NAMESPACE::shape_inference::InferShapes(model);
+  auto& outputs = model.graph().output();
+  for (auto& vi : outputs)  {
+    onnxTensorDescriptor outputDescriptor;
+    char* name = new char[vi.name().size() + 1];
+    strcpy(name, vi.name().c_str());
+    outputDescriptor.name = name;
+    outputDescriptor.memoryType = ONNXIFI_MEMORY_TYPE_CPU; //TODO: Expand memory types?
+    if (!vi.type().tensor_type().has_elem_type())  {
+      throw std::runtime_error("Non-static ModelProto: Output data type not found");
+    }
+    outputDescriptor.dataType = vi.type().tensor_type().elem_type();
+    outputDescriptor.dimensions = vi.type().tensor_type().shape().dim_size();
+    auto* shape = new uint64_t[outputDescriptor.dimensions];
```

Comment:
leak?

Reply:
same as above^

Reply:
I saw you are deleting them in the dtor. But it's still safe. For example, 4 lines down, you can `throw`. And you may do something else and this chunk of memory got leakage. 

I have a suggestion
You can do vector<uint64_t> shape, then fill in the shape and use `shape.data()` whenever you want a `uint64_t*`.

Reply:
Good point - Packaged data into DescriptorData struct with rvalue constructor; no more new/delete